SOLIDITY_STYLE_GUIDE.md

```diff
@@ -1851,3 +1853,57 @@ function processStaking(
     ...
 }
 ```
+## Security
+
+### Checks Effects Interactions
+
+Checks effects interactions pattern should be used for writing contract methods which make external calls. This pattern reduces the attack surface for malicious contracts trying to hijack control flow after an external call.
+
+When to use this pattern: 
+
+* It cannot be avoided to hand over control flow to an external entity.
+* You want to protect your functions against re-entrancy attacks.
+
+Details:
+
+1. **Checks**: If input is not acceptable then fail fast and fail hard. 
+
+    ```
+    require(someCondition);
+    
+    ```
+
+2. **Effects**: Optimistically update the contract to a new valid state assuming that interactions will be successful. This protects the contract from re-entrance and race conditions.
+
+    ```
+    balance[msg.sender] = 0;
+    
+    ```
+
+3. **Interactions**: External function calls, `.send()`, `.transfer()` and `.call()` should be done at the end of the method. The return value of external call should be asserted if possible. 
```

Comment:
There are 4 things: 
1. External function calls (other contract method call) 
2. `.send()`
3. `.transfer()`
4. `.call()`

I think you missed the first  point. is it not clear?

Reply:
@sarvesh-ost: Your response above provides the definition I was missing: that "external function call" means a call to a different contract's method:

> If my understanding is incorrect or my reading is correct, perhaps it would be helpful to define external function calls 🤔

I think it would be useful to add that to the guidance to make clear that the list *is* exhaustive, but that `send`, `transfer`, and `call` **are not considered external function calls**.

I am in favor of explicitly stating that these 4 items all concern the checks-effects-interactions pattern, but I am unable to find the distinction between the first item, external function calls, and the other 3 items that is made here in the reference source or explicitly in the Solidity documentation. I may have overlooked it—please point it out if I did 😰 Regardless, since it is a distinction that does not make a difference (even if only some of them are *external function calls*, they are all *external calls* and that's what concerns the pattern), I would vote for not making a distinction:

> 3. **Interactions**: External ~function~ calls, including `.send()`, `.transfer()` and `.call()`, should be done at the end of the method.

(strike "function" and add a comma after `.call()`)

But I will defer to your view on what you think will be the most useful to other devs.